### PR TITLE
Adding mocked tests for routes

### LINKED ISSUE #492: Increase code coverage of pkg/api/routes.go
Currently we have little unit tests in pkg/api/routes.go, it is mostly covered by integration tests. The integration tests may or may not hit some of the corner cases reliably, which results in different code coverage results when running the same tests on the same code.

We should implement more detailed tests to reproduce more of the errors which should be handled. Perhaps we could solve this by adding new tests which mock the storage, or other components we interact with in routes.go.

### PR DESCRIPTION
**Pull request by laurentiuNiculae**
**What type of PR is this?**

**Which issue does this PR fix**:
Issue #492
Closes #492
**What does this PR do / Why do we need it**:
Raise cod coverage for routes.go

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/routes_test.go [hunk 1]
```diff
@@ -0,0 +1,841 @@
...
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	. "github.com/smartystreets/goconvey/convey"
+	zerr "zotregistry.io/zot/errors"
+	"zotregistry.io/zot/pkg/api"
+	"zotregistry.io/zot/pkg/api/config"
+	"zotregistry.io/zot/pkg/api/constants"
+	"zotregistry.io/zot/pkg/test"
+)
+
+type ImageStoreMock struct {
```
Comment (rchincha) [inline pkg/api/routes_test.go:27]: call this "MockedImageStore" instead.

File: errors/errors.go [hunk 2]
```diff
@@ -53,4 +53,5 @@ var (
 	ErrRegistryNoContent       = errors.New("sync: could not find a Content that matches localRepo")
 	ErrSyncSignatureNotFound   = errors.New("sync: couldn't find any upstream notary/cosign signatures")
 	ErrSyncSignature           = errors.New("sync: couldn't get upstream notary/cosign signatures")
+	ErrUnexpectedError         = errors.New("error: unexpected error")
```
Comment (rchincha) [inline errors/errors.go:56]: Where is this actually used?

File: pkg/api/routes.go [hunk 3]
```diff
@@ -949,14 +949,6 @@ func (rh *RouteHandler) PatchBlobUpload(response http.ResponseWriter, request *h
 			return
 		}
 
-		contentRange := request.Header.Get("Content-Range")
```
Comment (rchincha) [inline pkg/api/routes.go:952]: This part got deleted? why?

Comment (laurentiuNiculae) [reply pkg/api/routes.go:952]: Because on line 937 we have this condition
`if request.Header.Get("Content-Length") == "" || request.Header.Get("Content-Range") == ""`

since the removed code is on the else branch, testing for Content-Range to be empty will always result false, so it's unreachable.

File: errors/errors.go (hunk 2 above)
Comment (laurentiuNiculae) [reply errors/errors.go:56]: I use it just in tests to be more expressive on the meaning of the error I test for.

File: pkg/api/routes.go (hunk 3 above)
Comment (peusebiu) [reply pkg/api/routes.go:958]: maybe there should be an AND instead of OR?

Comment (laurentiuNiculae) [reply pkg/api/routes.go]: I have tested this change and it breaks Zot.
`Summarizing 2 Failures:

[Fail] OCI Distribution Conformance Tests Push Blob Upload Streamed [It] PATCH request with blob in body should yield 202 response 
/go/src/github.com/opencontainers/distribution-spec/conformance/02_push_test.go:35

[Panic!] OCI Distribution Conformance Tests Push Blob Upload Streamed [It] PUT request to session URL with digest should yield 201 response 
/usr/local/go/src/runtime/panic.go:199
`

Comment (laurentiuNiculae) [reply pkg/api/routes.go]: I think the initial logic is good if it's compliant with the OCI tests. They seem quite exhaustive. I propose to just remove the redundant `if`.

File: errors/errors.go (hunk 2 above)
Comment (rchincha) [reply errors/errors.go:56]: Then move this error into the tests itself. It doesn't belong here.

File: pkg/api/routes.go (hunk 3 above)
Comment (rchincha) [reply pkg/api/routes.go:958]: https://github.com/project-zot/zot/pull/527

Refactored like so ^. This is what this check really means.

Comment (rchincha) [reply pkg/api/routes.go:958]: https://github.com/project-zot/zot/runs/6375017852?check_suite_focus=true
and conformance passes ^

File: errors/errors.go (hunk 2 above)
Comment (rchincha) [reply errors/errors.go:56]: maybe create a new errors/test_errors.go and move it there.

File: pkg/api/routes.go (hunk 3 above)
Comment (rchincha) [reply pkg/api/routes.go:958]: https://pkg.go.dev/net/http#Header.Set
^ Is it possible to set the key but empty value in this case?

Comment (laurentiuNiculae) [reply pkg/api/routes.go:958]: I am sorry but I don't understand what you mean by the last comment. I have written a small (but exhaustive) equivalent example to show that the line is unreachable.

```
package main

import (
	"fmt"
)

func foo(contentIsPresent, rangeIsPresent bool) {
	if !contentIsPresent || !rangeIsPresent { // equivalent to line 943
		fmt.Println("PutBlobChunkStreamed")
	} else {
		if !rangeIsPresent {                  // equivalent to line 959
			fmt.Println("Unreachable line")   
		}
	}
}

func main() {
	truthValues := [2]bool{true, false}

	for _, i := range truthValues {
		for _, j := range truthValues {
			foo(i, j)
		}
	}
}
```

File: errors/errors.go (hunk 2 above)
Comment (laurentiuNiculae) [reply errors/errors.go]: Ok! I have resolved this issue.

**Review by peusebiu — APPROVED** [review]
LGTM

**Review by rchincha — DISMISSED** [review]
lgtm

**Review by rchincha — DISMISSED** [review]
lgtm

**Review by rchincha — APPROVED** [review]
lgtm
